pkg/testcase/upgradenodereplacement.go

```diff
@@ -0,0 +1,570 @@
+		if joinErr != nil {
+			shared.LogLevel("error", "error joining agent: %w\n", joinErr)
+
+			return joinErr
+		}
+	}
+
+	return nil
+}
+
+func deleteAgents(a *aws.Client, c *factory.Cluster) error {
+	for _, i := range c.AgentIPs {
+		if deleteNodeErr := shared.DeleteClusterNode(i); deleteNodeErr != nil {
+			shared.LogLevel("error", "error deleting agent: %w\n", deleteNodeErr)
+
+			return deleteNodeErr
+		}
+		shared.LogLevel("info", "Node IP deleted from the cluster: %s\n", i)
+
+		err := a.DeleteInstance(i)
+		if err != nil {
+			return err
+		}
+		shared.LogLevel("info", "Instance IP deleted from cloud provider: %s\n", i)
+	}
+
+	return nil
+}
+
+func joinAgent(product, serverIp, token, version, selfExternalIp, selfPrivateIp string) error {
+	cmd, parseErr := parseJoinCmd(product, agent, serverIp, token, version, selfExternalIp, selfPrivateIp)
+	if parseErr != nil {
+		return shared.ReturnLogError("error parsing join commands: %w\n", parseErr)
+	}
+
+	if joinErr := joinNode(cmd, selfExternalIp); joinErr != nil {
+		return shared.ReturnLogError("error joining node: %w\n", joinErr)
+	}
+
+	return nil
+}
+
+func joinNode(cmd, ip string) error {
+	if cmd == "" {
+		return shared.ReturnLogError("cmd not sent\n")
+	}
+	if ip == "" {
+		return shared.ReturnLogError("server IP not sent\n")
+	}
+
+	res, err := shared.RunCommandOnNode(cmd, ip)
+	if err != nil {
+		return shared.ReturnLogError("error joining node: %w\n", err)
+	}
+	res = strings.TrimSpace(res)
+	if strings.Contains(res, "service failed") {
+		shared.LogLevel("error", "join node response: %s\n", res)
+
+		return shared.ReturnLogError("error joining node: %s\n", res)
+	}
+
+	delay := time.After(40 * time.Second)
+	// delay not meant to wait if node is joined, but rather to give time for all join process to complete under the hood
+	<-delay
+
+	return nil
+}
+
+func parseJoinCmd(product, nodetype, serverIp, token, version, selfExternalIp, selfPrivateIp string) (string, error) {
+	if nodetype != master && nodetype != agent {
+		return "", shared.ReturnLogError("unsupported nodetype: %s\n", nodetype)
+	}
+
+	var flags string
+	var instalMode string
+	if nodetype == master {
+		flags = fmt.Sprintf("'%s'", os.Getenv("server_flags"))
+	} else {
+		flags = fmt.Sprintf("'%s'", os.Getenv("worker_flags"))
+	}
+
+	if strings.HasPrefix(version, "v") {
+		instalMode = fmt.Sprintf("INSTALL_%s_VERSION", strings.ToUpper(product))
+	} else {
+		instalMode = fmt.Sprintf("INSTALL_%s_COMMIT", strings.ToUpper(product))
+	}
+
+	switch product {
+	case "k3s":
+		return parseK3sCmd(nodetype, serverIp, token, version, selfExternalIp, selfPrivateIp, instalMode, flags)
+	case "rke2":
+		return parseRke2Cmd(nodetype, serverIp, token, version, selfExternalIp, selfPrivateIp, instalMode, flags)
+	default:
+		return "", shared.ReturnLogError("unsupported product: %s\n", product)
+	}
+}
+
+func parseK3sCmd(nodetype, serverIp, token, version, selfExternalIp, selfPrivateIp, instalMode, flags string,
```

Comment:
Also the name is confusing, we are not parsing the command, right? we are building it and sending it back

Reply:
> Will it be too complicated to construct the command without duplication?

not complicated , but since they have different options for agent/master , it would be kind of mess and not clear to understand , so i intentionally decide to duplicate some efforts, in order of makes thing clearer and more maintanable here 




changed the func name!